File: app/frontend/src/api/dispersion.ts
/** dispersion requests and models; moved without changing payloads or responses. */
import { request } from './client'
import type { DeviceEvent } from './devices'

export type DispersionState = 'draft' | 'pre_excitation' | 'burn' | 'dark' | 'paused' | 'stopping' | 'completed' | 'failed' | 'stopped'

export type DispersionLine = {
  id: number
  task_id: number
  element: string
  wavelength_nm: number
  ccd_index: number
  expected_position: number | null
  located_position: number | null
  saved_position: number | null
  position_state: 'pending' | 'located' | 'saved'
  position_source: string | null
  position_frame_id: number | null
  adjustment_points: number
  created_at: string
  updated_at: string
}

export type DispersionCalibrationVersion = {
  id: number
  name: string
  version: number
  state: 'draft' | 'published' | 'superseded'
  calibration_id: number | null
  ccd_layout_id: number
  source_task_id: number | null
  coefficients: number[]
  residuals: Array<{ line_id: number; element: string; wavelength_nm: number; ccd_index: number; measured_position: number; predicted_position: number; residual_points: number }>
  wavelength_min: number
  wavelength_max: number
  residual_rms: number
  residual_max: number
  point_count: number
  residual_limit_points: number
  publishable: boolean
  created_at: string
}

export type DispersionTask = {
  id: number
  name: string
  status: DispersionState
  paused_from: string | null
  device_profile_id: number
  ccd_layout_id: number
  method_id: number | null
  method_version: number | null
  frame_count: number
  dark_frame_count: number
  pre_excitation_seconds: number
  sampling_period_seconds: number
  residual_limit_points: number
  ccd_indices: number[]
  condition: Record<string, unknown>
  adapter_session_id: string | null
  burn_frames_captured: number
  dark_frames_captured: number
  last_frame_index: number | null
  last_event: DeviceEvent | null
  failure_code: string | null
  failure_message: string | null
  lines: DispersionLine[]
  frame_summary: Array<{ phase: 'burn' | 'dark'; frame_count: number; last_frame_index: number }>
  calibrations: DispersionCalibrationVersion[]
  layout: { id: number; name: string; frame_count: number; ccds_per_frame: number; points_per_ccd: number; ccd_indices: number[] }
  profile: { id: number; name: string; transport: string }
  created_at: string
  updated_at: string
}

export type DispersionFrame = {
  id: number
  task_id: number
  phase: 'burn' | 'dark'
  frame_index: number
  ccd_index: number
  points: number[]
  sha256: string
  headers: number[]
  byte_length: number
  virtual_time_ms: number
  captured_at: string
}

export type DispersionOptions = {
  ccd_layouts: Array<{ id: number; name: string; frame_count: number; ccds_per_frame: number; points_per_ccd: number; point_width_um: number; ccd_indices: number[]; wavelength_min: number; wavelength_max: number }>
  calibration_versions: DispersionCalibrationVersion[]
  device_profiles: Array<{ id: number; name: string; transport: string; frame_count: number; ccds_per_frame: number; points_per_ccd: number; ccd_indices: number[] }>
  states: DispersionState[]
}

export const dispersionApi = {
  dispersionOptions: (token: string) => request<DispersionOptions>('/api/v1/dispersion/options', { headers: { Authorization: `Bearer ${token}` } }),
  dispersionTasks: (token: string) => request<DispersionTask[]>('/api/v1/dispersion/tasks', { headers: { Authorization: `Bearer ${token}` } }),
  createDispersionTask: (token: string, payload: Record<string, unknown>) => request<DispersionTask>('/api/v1/dispersion/tasks', { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify(payload) }),
  dispersionTask: (token: string, taskId: number) => request<DispersionTask>(`/api/v1/dispersion/tasks/${taskId}`, { headers: { Authorization: `Bearer ${token}` } }),
  startDispersionTask: (token: string, taskId: number) => request<DispersionTask>(`/api/v1/dispersion/tasks/${taskId}/start`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  stepDispersionTask: (token: string, taskId: number) => request<DispersionTask>(`/api/v1/dispersion/tasks/${taskId}/step`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  pauseDispersionTask: (token: string, taskId: number) => request<DispersionTask>(`/api/v1/dispersion/tasks/${taskId}/pause`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  resumeDispersionTask: (token: string, taskId: number) => request<DispersionTask>(`/api/v1/dispersion/tasks/${taskId}/resume`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  stopDispersionTask: (token: string, taskId: number) => request<DispersionTask>(`/api/v1/dispersion/tasks/${taskId}/stop`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  dispersionFrames: (token: string, taskId: number, ccdIndex?: number) => request<DispersionFrame[]>(`/api/v1/dispersion/tasks/${taskId}/frames${ccdIndex === undefined ? '' : `?ccd_index=${ccdIndex}`}`, { headers: { Authorization: `Bearer ${token}` } }),
  addDispersionLine: (token: string, taskId: number, payload: { element: string; wavelength_nm: number; ccd_index: number; actual_position?: number | null }) => request<DispersionLine>(`/api/v1/dispersion/tasks/${taskId}/lines`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify(payload) }),
  deleteDispersionLine: (token: string, taskId: number, lineId: number) => request<{ id: number; deleted: boolean }>(`/api/v1/dispersion/tasks/${taskId}/lines/${lineId}`, { method: 'DELETE', headers: { Authorization: `Bearer ${token}` } }),
  locateDispersionLine: (token: string, taskId: number, lineId: number) => request<DispersionLine>(`/api/v1/dispersion/tasks/${taskId}/lines/${lineId}/locate`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  locateAllDispersionLines: (token: string, taskId: number) => request<{ located: DispersionLine[]; errors: Array<Record<string, unknown>>; all_succeeded: boolean }>(`/api/v1/dispersion/tasks/${taskId}/lines/locate-all`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  moveDispersionLine: (token: string, taskId: number, lineId: number, direction: 'short' | 'long', steps = 1) => request<DispersionLine>(`/api/v1/dispersion/tasks/${taskId}/lines/${lineId}/move`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify({ direction, steps }) }),
  saveDispersionLinePosition: (token: string, taskId: number, lineId: number) => request<DispersionLine>(`/api/v1/dispersion/tasks/${taskId}/lines/${lineId}/position/save`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  restoreDispersionLinePosition: (token: string, taskId: number, lineId: number) => request<DispersionLine>(`/api/v1/dispersion/tasks/${taskId}/lines/${lineId}/position/restore`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  fitDispersionCalibration: (token: string, taskId: number, payload: { name?: string; degree?: number; residual_limit_points?: number }) => request<DispersionCalibrationVersion>(`/api/v1/dispersion/tasks/${taskId}/calibrations/fit`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify(payload) }),
  publishDispersionCalibration: (token: string, calibrationVersionId: number) => request<DispersionCalibrationVersion>(`/api/v1/dispersion/calibrations/${calibrationVersionId}/publish`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  bindDispersionCalibration: (token: string, calibrationVersionId: number, methodId: number, methodVersion?: number) => request<Record<string, unknown>>(`/api/v1/dispersion/calibrations/${calibrationVersionId}/bind`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify({ method_id: methodId, method_version: methodVersion }) }),
}

File: app/backend/api/dispersion.py
"""Dispersion API boundary; business services retain their existing behavior."""
from __future__ import annotations
from typing import Any
from fastapi import APIRouter, Depends, HTTPException, Query
from ..schemas.dispersion import (
    DispersionTaskCreate,
    DispersionLineInput,
    DispersionLineMoveRequest,
    DispersionCalibrationFitRequest,
    DispersionCalibrationBindRequest,
)
from ..auth import Session
from ..modules.dispersion import DispersionError
from ..runtime import Runtime
from .dependencies import get_runtime, require_permission

router = APIRouter()


def dispersion_error(exc: DispersionError) -> HTTPException:
    return HTTPException(status_code=exc.status_code, detail=exc.detail())


@router.get("/api/v1/dispersion/options", tags=["dispersion"])
def dispersion_options(_: Session = Depends(require_permission("dispersion.read")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    return runtime.dispersion_service().options()


@router.get("/api/v1/dispersion/tasks", tags=["dispersion"])
def list_dispersion_tasks(
    limit: int = Query(default=50, ge=1, le=200),
    _: Session = Depends(require_permission("dispersion.read")), *, runtime: Runtime = Depends(get_runtime)) -> list[dict[str, Any]]:
    return runtime.dispersion_service().list_tasks(limit)


@router.post("/api/v1/dispersion/tasks", status_code=201, tags=["dispersion"])
def create_dispersion_task(
    payload: DispersionTaskCreate,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().create_task(payload.model_dump(mode="json"), session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.get("/api/v1/dispersion/tasks/{task_id}", tags=["dispersion"])
def get_dispersion_task(
    task_id: int,
    _: Session = Depends(require_permission("dispersion.read")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().get_task(task_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/start", tags=["dispersion"])
def start_dispersion_task(
    task_id: int,
    session: Session = Depends(require_permission("dispersion.execute")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().start_task(task_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/step", tags=["dispersion"])
def step_dispersion_task(
    task_id: int,
    session: Session = Depends(require_permission("dispersion.execute")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().step_task(task_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/pause", tags=["dispersion"])
def pause_dispersion_task(
    task_id: int,
    session: Session = Depends(require_permission("dispersion.execute")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().pause_task(task_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/resume", tags=["dispersion"])
def resume_dispersion_task(
    task_id: int,
    session: Session = Depends(require_permission("dispersion.execute")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().resume_task(task_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/stop", tags=["dispersion"])
def stop_dispersion_task(
    task_id: int,
    session: Session = Depends(require_permission("dispersion.execute")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().stop_task(task_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.get("/api/v1/dispersion/tasks/{task_id}/frames", tags=["dispersion"])
def dispersion_task_frames(
    task_id: int,
    phase: str | None = Query(default=None),
    ccd_index: int | None = Query(default=None, ge=0, le=255),
    _: Session = Depends(require_permission("dispersion.read")), *, runtime: Runtime = Depends(get_runtime)) -> list[dict[str, Any]]:
    try:
        return runtime.dispersion_service().frames(task_id, phase=phase, ccd_index=ccd_index)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/lines", status_code=201, tags=["dispersion"])
def add_dispersion_line(
    task_id: int,
    payload: DispersionLineInput,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().add_line(task_id, payload.model_dump(mode="json"), session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.delete("/api/v1/dispersion/tasks/{task_id}/lines/{line_id}", tags=["dispersion"])
def delete_dispersion_line(
    task_id: int,
    line_id: int,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().delete_line(task_id, line_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/lines/locate-all", tags=["dispersion"])
def locate_all_dispersion_lines(
    task_id: int,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    return runtime.dispersion_service().locate_all(task_id, session.user_id)


@router.post("/api/v1/dispersion/tasks/{task_id}/lines/{line_id}/locate", tags=["dispersion"])
def locate_dispersion_line(
    task_id: int,
    line_id: int,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().locate_line(task_id, line_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/lines/{line_id}/move", tags=["dispersion"])
def move_dispersion_line(
    task_id: int,
    line_id: int,
    payload: DispersionLineMoveRequest,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().move_line(task_id, line_id, payload.direction, payload.steps, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/lines/{line_id}/position/save", tags=["dispersion"])
def save_dispersion_line_position(
    task_id: int,
    line_id: int,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().save_line_position(task_id, line_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/lines/{line_id}/position/restore", tags=["dispersion"])
def restore_dispersion_line_position(
    task_id: int,
    line_id: int,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().restore_line_position(task_id, line_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/tasks/{task_id}/calibrations/fit", status_code=201, tags=["dispersion"])
def fit_dispersion_calibration(
    task_id: int,
    payload: DispersionCalibrationFitRequest,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().fit_calibration(task_id, payload.model_dump(exclude_none=True), session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.get("/api/v1/dispersion/calibrations/{calibration_version_id}", tags=["dispersion"])
def get_dispersion_calibration(
    calibration_version_id: int,
    _: Session = Depends(require_permission("dispersion.read")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().calibration(calibration_version_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/calibrations/{calibration_version_id}/publish", tags=["dispersion"])
def publish_dispersion_calibration(
    calibration_version_id: int,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().publish_calibration(calibration_version_id, session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.post("/api/v1/dispersion/calibrations/{calibration_version_id}/bind", tags=["dispersion"])
def bind_dispersion_calibration(
    calibration_version_id: int,
    payload: DispersionCalibrationBindRequest,
    session: Session = Depends(require_permission("dispersion.write")), *, runtime: Runtime = Depends(get_runtime)) -> dict[str, Any]:
    try:
        return runtime.dispersion_service().bind_calibration(calibration_version_id, payload.model_dump(exclude_none=True), session.user_id)
    except DispersionError as exc:
        raise dispersion_error(exc) from exc


@router.get("/api/v1/dispersion/bindings", tags=["dispersion"])
def list_dispersion_bindings(
    method_id: int | None = Query(default=None, ge=1),
    _: Session = Depends(require_permission("dispersion.read")), *, runtime: Runtime = Depends(get_runtime)) -> list[dict[str, Any]]:
    return runtime.dispersion_service().bindings(method_id)
